### ISSUE TITLE
[FR] add the ability to set/change scalar field values when duplicating elements (like attributes)

### ISSUE THREAD

**Issue by Anubarak** [craftcms/cms#7726]
### Description

I would really like to set scalar (in best case even non scalar, but this seems to be trickier https://github.com/craftcms/cms/issues/6191) field values when duplicating elements.

For example duplicating an element and changing a field value requires me to save the element once more.

```PHP
$elements = Craft::$app->getElements();
$clone = $elements->duplicateElement($main, ['someAttribute' => 'new value']);
$clone->setFieldValue('myFieldValue', true);
$elements->saveElement($clone);
```

The following would be much cleaner imo

```PHP
$elements = Craft::$app->getElements();
$clone = $elements->duplicateElement($main, ['someAttribute' => 'new value'], ['myFieldValue' => true]);
```

or even 
```PHP
$clone = $elements->duplicateElement($main, ['someAttribute' => 'new value', 'customFields' => ['myFieldValue' => true]]);

via 
$fieldValues = ArrayHelper::remove($newAttributes, 'customFields', []);
```

**Comment by brandonkelly** [comment]
I’m actually pretty sure this is already (at least partially) supported, as custom fields are considered element attributes:

https://github.com/craftcms/cms/blob/05f280380028c062efa5acfb42d35259804c6240/src/base/Element.php#L1638-L1643

So you should be able to do:

```php
$clone = $elements->duplicateElement($main, [
    'someAttribute' => 'new value',
    'myFieldValue' => true,
]);
```

**Comment by Anubarak** [comment]
@brandonkelly yes indeed. I tought so too, but in the end of the day yii only executes `$behaviour->$key = $value` in the end of the day. It took me a while to figure it out but the `normalizeValue` isn't executed and even for "scalar" values that's sometimes required because some custom fields process these values in some way or another.
I debugged it

**Comment by brandonkelly** [comment]
Ah, right. OK, will think on the best way to solve this. Maybe we should stop allowing setting the CustomFieldBehavior attributes via `__set()`, and force those values through `setFieldValue()` instead.

**Comment by Anubarak** [comment]
What would be the downside to do the same like in your `__get` function and call the `setFieldValue` in case it is a custom field?

**Comment by brandonkelly** [comment]
That’s what I’m saying we should do. Not sure why it wasn’t like that to begin with, besides that it maybe didn’t feel necessary in the way `__get()` did. 🤷🏻

**Comment by brandonkelly** [comment]
Just made that change on the `3.7` branch.

**Comment by brandonkelly** [comment]
This one has been a bit of a trip. My original commit (5f5f389c854d73d483c291a6e525861775c23e50) caused custom fields to take precedence over native element setters, which shouldn’t be the case, so I move it down to only happen as a fallback if `parent::__set()` threw an exception (d6c0378b87c0eb3e452f4f3f13a04e89c460d2d2). However I just realized that wasn’t actually working as expected, because `parent::__set()` was already successfully setting the custom field property on the behavior (albeit w/out going through `setFieldValue()`). So the net result is that nothing changed in 3.7 related to setting custom field values.

I’ve just fixed that for the next release (f7235152d232af11c797de92bdb76049118c30c1). Going forward, `CustomFieldBehavior` classes can be configured to not allow their properties to be set via the owner’s `__set()` method:

https://github.com/craftcms/cms/blob/f7235152d232af11c797de92bdb76049118c30c1/src/behaviors/CustomFieldBehavior.php.template#L116-L120

And `craft\base\Element` is taking advantage of that:

https://github.com/craftcms/cms/blob/f7235152d232af11c797de92bdb76049118c30c1/src/base/Element.php#L1780-L1783

Which means that `yii\base\Component::__set()` will throw an `UnknownPropertyException` as expected, and `craft\base\Element::__set()` will successfully fall back to going through `setFieldValue()`:

https://github.com/craftcms/cms/blob/f7235152d232af11c797de92bdb76049118c30c1/src/base/Element.php#L1752-L1755

**Comment by brandonkelly** [comment]
Craft 3.7.21 is out now with those changes.
